firedrake/matrix_free/operators.py

```diff
@@ -161,27 +226,53 @@ def mult(self, mat, X, Y):
             v.copy(Y)
 
     def multTranspose(self, mat, Y, X):
-        # As for mult, just everything swapped round.
-        with self._y.dat.vec_wo as v:
-            Y.copy(v)
+        # EquationBC makes multTranspose different from mult.
+        #
+        # Decompose M^T into bundles of columns associated with
+        # the rows of M corresponding to cell, facet,
+        # edge, and vertice equations (if exist) and add up their
+        # contributions.
+
+        from firedrake.bcs import EquationBCSplit
+
+        iter_assemble_actionT = iter(self.list_assemble_actionT)
+
+        # accumulate values in self._xbc for convenience
+        self._xbc.dat.zero()
+
+        def multTransposePart(bc):
 
-        for bc in self.row_bcs:
-            bc.zero(self._y)
+            # TODO, can we avoid this copy, too?
+            with self._y.dat.vec_wo as v:
+                Y.copy(v)
 
-        self._assemble_actionT()
+            # zero columns associated with DirichletBCs/EquationBCs
+            for bbc in bc.bcs:
+                bbc.zero(self._y)
+
+            next(iter_assemble_actionT)()
+
+            self._xbc += self._x
+
+            # recursion
+            for bbc in bc.bcs:
+                if isinstance(bbc, EquationBCSplit):
+                    multTransposePart(bbc)
```

Comment:
There's a lot of hand-crafted recursion on all of these bc objects. Can we find a common pattern? It looks like you want `for bc in EquationBC` to iterate over all recursive bcs.

Reply:
This code is still quite impenetrable to me. I think it is doing something like:
```python
for action, obj in zip(assembly_actions, (self, all_eq_bcs)):
    copy Y -> self._y
    for bc in obj.bcs:
       bc.zero(self._y)
    action()
    self._xbc += self._x
```

So for every object that has a form, we're lifting the action of the form to the result. The copy has to happen each time, because ... (I don't know this). Can you write out the maths?

In any case, this crafted recursion suggests we should have convenience:

```python
eq_bcs = tuple(bc for bc in itertools.chain(*list_of_bcs) if isinstance(bc, EquationBCSplit))
```

Reply:
Thanks! I cleaned this up using `zip` and added some explanation of the algorithm in the comment.  It turned out that we actually can avoid many copies here.